conda_libmamba_solver/state.py

```diff
@@ -218,11 +219,40 @@ def _check_state(self):
         """
         Run some consistency checks to ensure configuration is solid.
         """
-        # Ensure configured pins match installed builds
-        for name, pin_spec in self.pinned.items():
-            installed = self.installed.get(name)
-            if installed and not pin_spec.match(installed):
-                raise SpecsConfigurationConflictError([installed], [pin_spec], self.prefix)
+        # Check we are not trying to remove things that are not installed
+        if self.is_removing:
+            not_installed = [
+                spec for name, spec in self.requested.items() if name not in self.installed
+            ]
+            if not_installed:
+                exc = PackagesNotFoundError(not_installed)
+                exc.allow_retry = False
+                raise exc
+
+        # Check if requested and pins overlap
+        # NOTE: This is a difference with respect to classic logic. classic
+        # allows pin overrides in the CLI, but we don't.
+        constraining_requests = {
+            spec.name for spec in self.requested.values() if not spec.is_name_only_spec
+        }
+        constraining_pins = {spec.name for spec in self.pinned.values()}
+        requested_and_pinned = constraining_requests.intersection(constraining_pins)
+        if requested_and_pinned:
+            # libmamba has two types of pins:
+            # - Pins with constraints: limit which versions of a package can be installed
+            # - Name-only pins: lock the package as installed; has no effect if not installed
+            #   Below we render name-only pins as their installed version when appropriate.
+            pinned_specs = [
+                (self.installed.get(name, pin) if pin.is_name_only_spec else pin)
+                for name, pin in sorted(self.pinned.items())
+            ]
+            exc = RequestedAndPinnedError(
+                requested_specs=sorted(self.requested.values(), key=lambda x: x.name),
+                pinned_specs=pinned_specs,
+                prefix=self.prefix,
+            )
+            exc.allow_retry = False
+            raise exc
```

Comment:
We moved these clauses here (from `SolverOutputState`) because at this point we haven't initialized the `index` (which means downloading repodata and parsing it, with the implied IO overheads). Since we don't need it to compute these errors, we can exit even earlier and happier users. Imagine waiting precious seconds downloading repodata only to realize you mistyped something in the CLI. Now it's mostly instantaneous.